nodejs/toolchain.bzl

```diff
@@ -25,6 +25,31 @@ May be empty if the target_tool_path points to a locally installed node binary."
     },
 )
 
+def node_toolchain_providers(ctx, toolchain_info):
+    """Convert the toolchain_info struct to Providers returned by the toolchain.
+
+       This is exposed as part of the workaround for https://github.com/bazelbuild/bazel/issues/14009
+    """
+    return [
+        DefaultInfo(
+            files = depset(toolchain_info.tool_files),
+            runfiles = ctx.runfiles(files = toolchain_info.tool_files),
+        ),
+        platform_common.ToolchainInfo(
+            nodeinfo = NodeInfo(
+                target_tool_path = toolchain_info.target_tool_path,
+                tool_files = toolchain_info.tool_files,
+            ),
+        ),
+        # Make the $(NODE_PATH) variable available in places like genrules.
+        # See https://docs.bazel.build/versions/main/be/make-variables.html#custom_variables
+        # Note that genrule seems to have a bug: this only works if the node_toolchain target
+        # itself is used by the toolchains attribute of genrule, not the toolchain_type.
+        platform_common.TemplateVariableInfo({
```

Comment:
Is it worth having the actual `node_toolchain` rule provide this (and any future templated variables), and having this function just fetch and forward the provider?

Reply:
"fetch and forward the provider" is what I could not figure out. In the impl of the `resolved_toolchain` rule I looked around at the 
`dir(ctx.toolchains["@rules_nodejs//nodejs:toolchain_type"])`
 and
`dir(ctx.toolchains["@rules_nodejs//nodejs:toolchain_type"].nodeinfo)`
and tried various ways to access the providers and forward them, but failed.

Reply:
Right, this was unclear and I apologize.

The result type from `ctx.toolchains[anything]` is a `ToolchainInfo` provider. In your case, this has a `nodeinfo` field which stores your `NodeInfo` provider.

My recommendation is that in the `node_toolchain` rule, you create both the `NodeInfo` and `TemplateVariableInfo` providers, and put them into the `ToolchainInfo` with different keys (`nodeinfo` and `template_variables` would work).

Then, in `resolved_toolchain`, you get back the `ToolchainInfo` (by calling `ctx.toolchains["@rules_nodejs//nodejs:toolchain_type"]`), and have `resolved_toolchain` return this:

```
    toolchainInfo = ctx.toolchains["@rules_nodejs//nodejs:toolchain_type"]
    return [
      toolchain_info,
      toolchain_info.nodeinfo,
      toolchain_info.template_variables,
    ]
```

That way, all three providers from `node_toolchain` are re-exported by `resolved_toolchain`.

Reply:
This is exactly the pattern I implemented in https://github.com/bazelbuild/rules_perl/pull/27 in case it's easy to crib :)

Reply:
you're very kind to give such hand-holding and I don't expect an apology at all! very appreciative, I don't know where you find the time to be so helpful 🙏🏻